source/Nevermore/Advanced/DeadlockAwareLock.cs

```diff
@@ -12,34 +14,53 @@ namespace Nevermore.Advanced
     ///     going to be a deadlock. In other words: very few false positives; probably some false negatives; better than
     ///     nothing.
     /// </summary>
-    public class DeadlockAwareLock : SemaphoreSlim
+    public class DeadlockAwareLock : IDisposable
     {
+        static readonly ILog Log = LogProvider.For<DeadlockAwareLock>();
+        
+        readonly SemaphoreSlim semaphore = new(1, 1);
+        
         int? taskWhichHasAcquiredLock;
         int? threadWhichHasAcquiredLock;
 
-        public DeadlockAwareLock() : base(1, 1)
+        readonly bool logOnConcurrentExecution;
+
+        public DeadlockAwareLock(bool logOnConcurrentExecution)
         {
+            this.logOnConcurrentExecution = logOnConcurrentExecution;
         }
 
-        public new void Wait()
+        public void Wait()
         {
             AssertNoDeadlock();
-            base.Wait();
+
+            if (logOnConcurrentExecution && semaphore.CurrentCount == 0)
```

Comment:
TIL about CurrentCount. Nicer than a Wait(0)

Reply:
Yeah, it is pretty chill, though I might add a comment there (since it's not always clear that `SemaphoreSlim` counts down instead of up)!